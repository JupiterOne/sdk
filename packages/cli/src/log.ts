import chalk from 'chalk';

/* eslint-disable no-console */
export function debug(msg: string) {
  console.log(`${chalk.gray(msg)}`);
}

export function info(msg: string) {
  console.log(`${chalk.white(msg)}`);
}

export function warn(msg: string) {
  console.log(`${chalk.yellow(msg)}`);
}

export function error(msg: string) {
  console.log(`${chalk.red(msg)}`);
}
