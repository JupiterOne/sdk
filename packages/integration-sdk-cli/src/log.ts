import chalk from 'chalk';

import {
  ExecuteIntegrationResult,
  processDeclaredTypesDiff,
} from '@jupiterone/integration-sdk-runtime';
import {
  IntegrationStepResult,
  StepResultStatus,
  SynchronizationJob,
  Metric,
} from '@jupiterone/integration-sdk-core';

export function debug(msg: string) {
  console.log(`${chalk.gray(msg)}`);
}

export function info(msg: string) {
  console.log(`${chalk.white(msg)}`);
}

export function warn(msg: string) {
  console.log(`${chalk.yellow(msg)}`);
}

export function error(msg: string) {
  console.log(`${chalk.red(msg)}`);
}

export function displaySynchronizationResults(job: SynchronizationJob) {
  info('\nSynchronization results:\n');
  info(`Synchronization job status: ${chalk.cyan(job.status)}`);
  info(`Entities uploaded: ${chalk.cyan(job.numEntitiesUploaded)}`);
  info(`Relationships uploaded: ${chalk.cyan(job.numRelationshipsUploaded)}`);
}

export function displayExecutionResults(results: ExecuteIntegrationResult) {
  info('\nResults:\n');

  let undeclaredTypesDetected: boolean = false;

  processDeclaredTypesDiff(results, (step, undeclaredTypes) => {
    logStepStatus(step);

    if (step.status === StepResultStatus.SUCCESS && undeclaredTypes.length) {
      undeclaredTypesDetected = true;
      const undeclaredTypesList = undeclaredTypes.map((type) => {
        return `\n  - ${type}`;
      });
      warn(
        `The following types were encountered but are not declared in the step's "types" field:${undeclaredTypesList}`,
      );
    }
  });

  if (undeclaredTypesDetected) {
    warn(
      `\nUndeclared types were detected!
To ensure that integration failures do not cause accidental data loss,
please ensure that all known entity and relationship types
collected by a step are declared in the step's "types" field.`,
    );
  }

  const { partialDatasets } = results.metadata;

  if (partialDatasets.types.length) {
    warn('\nThe following datasets were marked as partial:');
    info(partialDatasets.types.map((type) => `  - ${type}`).join('\n'));
  }
}

export function displayMetrics(metrics: Metric[]) {
  metrics.forEach((metric) => {
    switch (metric.unit) {
      case 'Milliseconds':
        info(`Metric "${metric.name}" = ${metric.value}ms`);
        break;
    }
  });
}

function logStepStatus(stepResult: IntegrationStepResult) {
  const stepPrefix = chalk.white(
    `Step "${chalk.cyan(stepResult.id)}" completed with status:`,
  );
  const statusText = getStepStatusText(stepResult.status);
  console.log(`${stepPrefix} ${statusText}`);
}

function getStepStatusText(status: StepResultStatus) {
  switch (status) {
    case StepResultStatus.SUCCESS:
    case StepResultStatus.CACHED:
      return chalk.green(status);
    case StepResultStatus.FAILURE:
      return chalk.red(status);
    case StepResultStatus.PARTIAL_SUCCESS_DUE_TO_DEPENDENCY_FAILURE:
      return chalk.yellow(status);
    case StepResultStatus.DISABLED:
      return chalk.gray(status);
  }
}
