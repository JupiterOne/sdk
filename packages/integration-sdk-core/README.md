# @jupiterone/integration-sdk-core

_NOTE:_ This project is currently under development and the API interface is not
stable. Use at your own risk.

This package contains core utilties and types that integration developers will
interface with.

## Installation

```
npm install @jupiterone/integration-sdk-core

# or

npm install @jupiterone/integration-sdk-core
```
