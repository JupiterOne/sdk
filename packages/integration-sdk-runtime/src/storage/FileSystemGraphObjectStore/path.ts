import { CollectionType } from '@jupiterone/integration-sdk-core';
import path from 'path';

interface BuildObjectCollectionFilePathInput {
  storageDirectoryPath: string;
  collectionType: CollectionType;
  filename: string;
}

export function buildObjectCollectionFilePath({
  storageDirectoryPath,
  collectionType,
  filename,
}: BuildObjectCollectionFilePathInput) {
  return path.join('graph', storageDirectoryPath, collectionType, filename);
}

interface BuildIndexFilePathInput extends BuildIndexDirectoryPathInput {
  filename: string;
}

export function buildIndexFilePath({
  collectionType,
  type,
  filename,
}: BuildIndexFilePathInput) {
  return path.join(buildIndexDirectoryPath({ collectionType, type }), filename);
}

interface BuildIndexDirectoryPathInput {
  collectionType: CollectionType;
  type: string;
}

export function buildIndexDirectoryPath({
  collectionType,
  type,
}: BuildIndexDirectoryPathInput) {
  return path.join('index', collectionType, type);
}
