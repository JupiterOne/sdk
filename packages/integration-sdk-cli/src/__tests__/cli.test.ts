import * as nodeFs from 'fs';
import path from 'path';

import { StepResultStatus } from '@jupiterone/integration-sdk-core';
import {
  getProjectDirectoryPath,
  loadProjectStructure,
} from '@jupiterone/integration-sdk-private-test-utils';

import { createCli } from '../index';
import * as log from '../log';

const fs = nodeFs.promises;

jest.mock('../log');

function getDocumentationFilePath(fixtureName: string) {
  return path.join(getProjectDirectoryPath(fixtureName), 'docs/jupiterone.md');
}

async function documentCommandSnapshotTest(fixtureName: string) {
  loadProjectStructure(fixtureName);

  const writeFileSpy = jest
    .spyOn(fs, 'writeFile')
    .mockResolvedValueOnce(Promise.resolve());

  await createCli().parseAsync(['node', 'j1-integration', 'document']);

  expect(writeFileSpy).toHaveBeenCalledTimes(1);
  expect(writeFileSpy).toHaveBeenCalledWith(
    getDocumentationFilePath(fixtureName),
    expect.any(String),
    {
      encoding: 'utf-8',
    },
  );

  expect(writeFileSpy.mock.calls[0][1]).toMatchSnapshot();
}

afterEach(() => {
  delete process.env.MY_CONFIG;
});

describe('collect', () => {
  beforeEach(() => {
    loadProjectStructure('instanceWithDependentSteps');
  });

  test('loads the integration, executes it, and logs the result', async () => {
    await createCli().parseAsync(['node', 'j1-integration', 'collect']);

    expect(log.displayExecutionResults).toHaveBeenCalledTimes(1);
    expect(log.displayExecutionResults).toHaveBeenCalledWith({
      integrationStepResults: [
        {
          id: 'fetch-accounts',
          name: 'Fetch Accounts',
          declaredTypes: ['my_account'],
          partialTypes: [],
          encounteredTypes: [],
          status: StepResultStatus.SUCCESS,
        },
        {
          id: 'fetch-groups',
          dependsOn: ['fetch-accounts'],
          name: 'Fetch Groups',
          declaredTypes: ['my_groups'],
          partialTypes: [],
          encounteredTypes: [],
          status: StepResultStatus.SUCCESS,
        },
        {
          id: 'fetch-users',
          name: 'Fetch Users',
          declaredTypes: ['my_user'],
          partialTypes: [],
          encounteredTypes: [],
          status: StepResultStatus.SUCCESS,
        },
      ],
      metadata: {
        partialDatasets: {
          types: [],
        },
      },
    });
  });

  describe('schema validation', () => {
    beforeEach(() => {
      loadProjectStructure('instanceWithNonValidatingSteps');
      delete process.env.ENABLE_GRAPH_OBJECT_SCHEMA_VALIDATION;
    });

    test('step should fail if enableSchemaValidation = true', async () => {
      await createCli().parseAsync(['node', 'j1-integration', 'collect']);

      expect(log.displayExecutionResults).toHaveBeenCalledTimes(1);
      expect(log.displayExecutionResults).toHaveBeenCalledWith({
        integrationStepResults: [
          {
            id: 'fetch-users',
            name: 'Fetch Users',
            declaredTypes: ['my_user'],
            partialTypes: [],
            encounteredTypes: [],
            status: StepResultStatus.FAILURE,
          },
        ],
        metadata: {
          partialDatasets: {
            types: ['my_user'],
          },
        },
      });
    });

    test('step should pass if enableSchemaValidation = false', async () => {
      await createCli().parseAsync([
        'node',
        'j1-integration',
        'collect',
        '--disable-schema-validation',
      ]);

      expect(log.displayExecutionResults).toHaveBeenCalledTimes(1);
      expect(log.displayExecutionResults).toHaveBeenCalledWith({
        integrationStepResults: [
          {
            id: 'fetch-users',
            name: 'Fetch Users',
            declaredTypes: ['my_user'],
            partialTypes: [],
            encounteredTypes: ['my_user'],
            status: StepResultStatus.SUCCESS,
          },
        ],
        metadata: {
          partialDatasets: {
            types: [],
          },
        },
      });
    });
  });

  describe('-s or --step option', () => {
    test('will only run the steps provided if the -s / --steps option is passed in', async () => {
      await createCli().parseAsync([
        'node',
        'j1-integration',
        'collect',
        '--step',
        'fetch-users',
      ]);

      expect(log.displayExecutionResults).toHaveBeenCalledTimes(1);
      expect(log.displayExecutionResults).toHaveBeenCalledWith({
        integrationStepResults: [
          {
            id: 'fetch-accounts',
            name: 'Fetch Accounts',
            declaredTypes: ['my_account'],
            partialTypes: [],
            encounteredTypes: [],
            status: StepResultStatus.DISABLED,
          },
          {
            id: 'fetch-groups',
            dependsOn: ['fetch-accounts'],
            name: 'Fetch Groups',
            declaredTypes: ['my_groups'],
            partialTypes: [],
            encounteredTypes: [],
            status: StepResultStatus.DISABLED,
          },
          {
            id: 'fetch-users',
            name: 'Fetch Users',
            declaredTypes: ['my_user'],
            partialTypes: [],
            encounteredTypes: [],
            status: StepResultStatus.SUCCESS,
          },
        ],
        metadata: {
          partialDatasets: {
            types: [],
          },
        },
      });
    });

    test('will allow filtering to multiple steps if the -s or --steps option has multiple comma separated values', async () => {
      await createCli().parseAsync([
        'node',
        'j1-integration',
        'collect',
        '--step',
        'fetch-users,fetch-accounts',
      ]);

      // fetch-users and fetch-accounts have no dependents and should be the only
      // ones enabled.
      expect(log.displayExecutionResults).toHaveBeenCalledTimes(1);
      expect(log.displayExecutionResults).toHaveBeenCalledWith({
        integrationStepResults: [
          {
            id: 'fetch-accounts',
            name: 'Fetch Accounts',
            declaredTypes: ['my_account'],
            partialTypes: [],
            encounteredTypes: [],
            status: StepResultStatus.SUCCESS,
          },
          {
            id: 'fetch-groups',
            dependsOn: ['fetch-accounts'],
            name: 'Fetch Groups',
            declaredTypes: ['my_groups'],
            partialTypes: [],
            encounteredTypes: [],
            status: StepResultStatus.DISABLED,
          },
          {
            id: 'fetch-users',
            name: 'Fetch Users',
            declaredTypes: ['my_user'],
            partialTypes: [],
            encounteredTypes: [],
            status: StepResultStatus.SUCCESS,
          },
        ],
        metadata: {
          partialDatasets: {
            types: [],
          },
        },
      });
    });

    test('will allow filtering to multiple steps if the -s or --steps option is provided multiple times', async () => {
      await createCli().parseAsync([
        'node',
        'j1-integration',
        'collect',
        '--step',
        'fetch-users',
        '--step',
        'fetch-accounts',
      ]);

      // fetch-users and fetch-accounts have no dependents and should be the only
      // ones enabled.
      expect(log.displayExecutionResults).toHaveBeenCalledTimes(1);
      expect(log.displayExecutionResults).toHaveBeenCalledWith({
        integrationStepResults: [
          {
            id: 'fetch-accounts',
            name: 'Fetch Accounts',
            declaredTypes: ['my_account'],
            partialTypes: [],
            encounteredTypes: [],
            status: StepResultStatus.SUCCESS,
          },
          {
            id: 'fetch-groups',
            dependsOn: ['fetch-accounts'],
            name: 'Fetch Groups',
            declaredTypes: ['my_groups'],
            partialTypes: [],
            encounteredTypes: [],
            status: StepResultStatus.DISABLED,
          },
          {
            id: 'fetch-users',
            name: 'Fetch Users',
            declaredTypes: ['my_user'],
            partialTypes: [],
            encounteredTypes: [],
            status: StepResultStatus.SUCCESS,
          },
        ],
        metadata: {
          partialDatasets: {
            types: [],
          },
        },
      });
    });

    test('will run dependent steps of steps passed in', async () => {
      await createCli().parseAsync([
        'node',
        'j1-integration',
        'collect',
        '--step',
        'fetch-groups',
      ]);
      // fetch-groups depends on fetch-accounts so they should both
      // be enabled.
      expect(log.displayExecutionResults).toHaveBeenCalledTimes(1);
      expect(log.displayExecutionResults).toHaveBeenCalledWith({
        integrationStepResults: [
          {
            id: 'fetch-accounts',
            name: 'Fetch Accounts',
            declaredTypes: ['my_account'],
            partialTypes: [],
            encounteredTypes: [],
            status: StepResultStatus.SUCCESS,
          },
          {
            id: 'fetch-groups',
            dependsOn: ['fetch-accounts'],
            name: 'Fetch Groups',
            declaredTypes: ['my_groups'],
            partialTypes: [],
            encounteredTypes: [],
            status: StepResultStatus.SUCCESS,
          },
          {
            id: 'fetch-users',
            name: 'Fetch Users',
            declaredTypes: ['my_user'],
            partialTypes: [],
            encounteredTypes: [],
            status: StepResultStatus.DISABLED,
          },
        ],
        metadata: {
          partialDatasets: {
            types: [],
          },
        },
      });
    });
  });
});

describe('visualize', () => {
  let htmlFileLocation;
  beforeEach(() => {
    loadProjectStructure('typeScriptVisualizeProject');
    htmlFileLocation = path.resolve(
      process.cwd(),
      'custom-integration',
      'index.html',
    );
  });

  test('writes graph to html file', async () => {
    await createCli().parseAsync([
      'node',
      'j1-integration',
      'visualize',
      '--data-dir',
      path.resolve(process.cwd(), 'custom-integration'),
      '--output-file',
      htmlFileLocation,
    ]);

    const content = await fs.readFile(htmlFileLocation, 'utf8');

    const nodesRegex = /var nodes = new vis.DataSet\(\[{.*},{.*}\]\);/g;
    const edgesRegex = /var edges = new vis.DataSet\(\[{.*}\]\);/g;

    expect(log.info).toHaveBeenCalledWith(
      `Visualize graph here: ${htmlFileLocation}`,
    );
    expect(content).toEqual(expect.stringMatching(nodesRegex));
    expect(content).toEqual(expect.stringMatching(edgesRegex));
  });
});

describe('collect/visualize integration', () => {
  let htmlFileLocation;
  beforeEach(() => {
    loadProjectStructure('typeScriptIntegrationProject');

    htmlFileLocation = path.resolve(
      process.cwd(),
      '.j1-integration',
      'index.html',
    );
  });

  test('creates graph based on integration data', async () => {
    await createCli().parseAsync(['node', 'j1-integration', 'collect']);
    await createCli().parseAsync(['node', 'j1-integration', 'visualize']);

    const content = await fs.readFile(htmlFileLocation, 'utf8');

    const nodesRegex = /var nodes = new vis.DataSet\(\[{.*},{.*}\]\);/g;
    const edgesRegex = /var edges = new vis.DataSet\(\[{.*}\]\);/g;

    expect(log.info).toHaveBeenCalledWith(
      `Visualize graph here: ${htmlFileLocation}`,
    );
    expect(content).toEqual(expect.stringMatching(nodesRegex));
    expect(content).toEqual(expect.stringMatching(edgesRegex));
  });
});

describe('document', () => {
  test('loads the integration with entities and relationships and writes documentation results', async () => {
    await documentCommandSnapshotTest('docsInstanceWithRelationships');
  });

  test('loads the integration without relationships and writes documentation results', async () => {
    await documentCommandSnapshotTest(
      'docsInstanceWithDependentStepsNoRelationships',
    );
  });

  test('loads the integration without entities and writes documentation results', async () => {
    await documentCommandSnapshotTest('docsInstanceNoEntities');
  });

  test('loads the integration with array of entity classes and writes document', async () => {
    await documentCommandSnapshotTest('docsInstanceArrayEntityClasses');
  });

  test('loads the integration and writes documentation when documentation markers already exist', async () => {
    await documentCommandSnapshotTest('docsWithExistingGeneratedDocs');
  });

  test('handles duplicate entity _type ingested in multiple steps', async () => {
    await documentCommandSnapshotTest('docsInstanceDuplicateEntityTypes');
  });

  test('handles duplicate relationship _type ingested in multiple steps', async () => {
    await documentCommandSnapshotTest('docsInstanceDuplicateRelationshipTypes');
  });

  test('alphabetizes "entities" metadata', async () => {
    await documentCommandSnapshotTest('docsInstanceEntitiesAlphabetize');
  });

  test('alphabetizes "relationships" metadata', async () => {
    await documentCommandSnapshotTest('docsInstanceRelationshipsAlphabetize');
  });

  test('loads the integration with entity and mapped relationship and writes documentation results', async () => {
    await documentCommandSnapshotTest('docsInstanceWithMappedRelationships');
  });

  test('should allow passing a file path for the generated documentation', async () => {
    loadProjectStructure('docsInstanceCustomDocLoc');

    const writeFileSpy = jest
      .spyOn(fs, 'writeFile')
      .mockResolvedValueOnce(Promise.resolve());

    const customDocumentationFilePath = path.join(
      getProjectDirectoryPath('docsInstanceCustomDocLoc'),
      'custom-docs/custom-jupiterone.md',
    );

    await createCli().parseAsync([
      'node',
      'j1-integration',
      'document',
      '-o',
      'custom-docs/custom-jupiterone.md',
    ]);

    expect(writeFileSpy).toHaveBeenCalledTimes(1);
    expect(writeFileSpy).toHaveBeenCalledWith(
      customDocumentationFilePath,
      expect.any(String),
      {
        encoding: 'utf-8',
      },
    );

    expect(writeFileSpy.mock.calls[0][1]).toMatchSnapshot();
  });

  test('loads the integration without entities or relationships and does not write a file', async () => {
    loadProjectStructure('docsInstanceNoEntitiesNoRelationships');

    const writeFileSpy = jest
      .spyOn(fs, 'writeFile')
      .mockResolvedValueOnce(Promise.resolve());

    await createCli().parseAsync(['node', 'j1-integration', 'document']);
    expect(writeFileSpy).toHaveBeenCalledTimes(0);
    expect(log.info).toHaveBeenCalledWith(
      `No entities or relationships found to generate documentation for. Exiting.`,
    );
  });

  test('should throw error when an existing doc does not exist', async () => {
    loadProjectStructure('docsWithoutExistingDoc');

    const writeFileSpy = jest
      .spyOn(fs, 'writeFile')
      .mockResolvedValueOnce(Promise.resolve());

    const documentationFilePath = getDocumentationFilePath(
      'docsWithoutExistingDoc',
    );
    const expectedErrorMessageThrown = `ENOENT: no such file or directory, open '${documentationFilePath}'`;

    await expect(
      createCli().parseAsync(['node', 'j1-integration', 'document']),
    ).rejects.toThrowError(expectedErrorMessageThrown);

    expect(log.error).toHaveBeenCalledTimes(1);
    expect(log.error).toHaveBeenCalledWith(
      `Error loading documentation file from path (path=${documentationFilePath}, err=${expectedErrorMessageThrown})`,
    );
    expect(writeFileSpy).toHaveBeenCalledTimes(0);
  });
});
