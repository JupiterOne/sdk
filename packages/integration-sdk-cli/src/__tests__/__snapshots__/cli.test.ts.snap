// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`document alphabetizes "entities" metadata 1`] = `
"# Integration with JupiterOne

## Setup

In this section, please provide details about how to set up the integration with
JupiterOne. This may require provisioning some resources on the provider's side
(perhaps a role, app, or api key) and passing information over to JupiterOne.


<!-- {J1_DOCUMENTATION_MARKER_START} -->
<!--
********************************************************************************
NOTE: ALL OF THE FOLLOWING DOCUMENTATION IS GENERATED USING THE
\\"j1-integration document\\" COMMAND. DO NOT EDIT BY HAND! PLEASE SEE THE DEVELOPER
DOCUMENTATION FOR USAGE INFORMATION:

https://github.com/JupiterOne/sdk/blob/main/docs/integrations/development.md
********************************************************************************
-->

## Data Model

### Entities

The following entities are created:

| Resources   | Entity \`_type\` | Entity \`_class\` |
| ----------- | -------------- | --------------- |
| The Group   | \`my_group\`     | \`Group\`         |
| The Group 2 | \`my_group_2\`   | \`Group\`         |
| The User    | \`my_user\`      | \`User\`          |
| The User 2  | \`my_user_2\`    | \`User\`          |

<!--
********************************************************************************
END OF GENERATED DOCUMENTATION AFTER BELOW MARKER
********************************************************************************
-->
<!-- {J1_DOCUMENTATION_MARKER_END} -->"
`;

exports[`document alphabetizes "relationships" metadata 1`] = `
"# Integration with JupiterOne

## Setup

In this section, please provide details about how to set up the integration with
JupiterOne. This may require provisioning some resources on the provider's side
(perhaps a role, app, or api key) and passing information over to JupiterOne.


<!-- {J1_DOCUMENTATION_MARKER_START} -->
<!--
********************************************************************************
NOTE: ALL OF THE FOLLOWING DOCUMENTATION IS GENERATED USING THE
\\"j1-integration document\\" COMMAND. DO NOT EDIT BY HAND! PLEASE SEE THE DEVELOPER
DOCUMENTATION FOR USAGE INFORMATION:

https://github.com/JupiterOne/sdk/blob/main/docs/integrations/development.md
********************************************************************************
-->

## Data Model

### Relationships

The following relationships are created/mapped:

| Source Entity \`_type\` | Relationship \`_class\` | Target Entity \`_type\` |
| --------------------- | --------------------- | --------------------- |
| \`the_root\`            | **HAS**               | \`my_account\`          |
| \`the_root\`            | **HAS**               | \`my_account_1\`        |
| \`the_root\`            | **HAS**               | \`my_account_2\`        |

<!--
********************************************************************************
END OF GENERATED DOCUMENTATION AFTER BELOW MARKER
********************************************************************************
-->
<!-- {J1_DOCUMENTATION_MARKER_END} -->"
`;

exports[`document handles duplicate entity _type ingested in multiple steps 1`] = `
"# Integration with JupiterOne

## Setup

In this section, please provide details about how to set up the integration with
JupiterOne. This may require provisioning some resources on the provider's side
(perhaps a role, app, or api key) and passing information over to JupiterOne.


<!-- {J1_DOCUMENTATION_MARKER_START} -->
<!--
********************************************************************************
NOTE: ALL OF THE FOLLOWING DOCUMENTATION IS GENERATED USING THE
\\"j1-integration document\\" COMMAND. DO NOT EDIT BY HAND! PLEASE SEE THE DEVELOPER
DOCUMENTATION FOR USAGE INFORMATION:

https://github.com/JupiterOne/sdk/blob/main/docs/integrations/development.md
********************************************************************************
-->

## Data Model

### Entities

The following entities are created:

| Resources | Entity \`_type\` | Entity \`_class\` |
| --------- | -------------- | --------------- |
| The Group | \`my_group\`     | \`Group\`         |

<!--
********************************************************************************
END OF GENERATED DOCUMENTATION AFTER BELOW MARKER
********************************************************************************
-->
<!-- {J1_DOCUMENTATION_MARKER_END} -->"
`;

exports[`document handles duplicate relationship _type ingested in multiple steps 1`] = `
"# Integration with JupiterOne

## Setup

In this section, please provide details about how to set up the integration with
JupiterOne. This may require provisioning some resources on the provider's side
(perhaps a role, app, or api key) and passing information over to JupiterOne.


<!-- {J1_DOCUMENTATION_MARKER_START} -->
<!--
********************************************************************************
NOTE: ALL OF THE FOLLOWING DOCUMENTATION IS GENERATED USING THE
\\"j1-integration document\\" COMMAND. DO NOT EDIT BY HAND! PLEASE SEE THE DEVELOPER
DOCUMENTATION FOR USAGE INFORMATION:

https://github.com/JupiterOne/sdk/blob/main/docs/integrations/development.md
********************************************************************************
-->

## Data Model

### Relationships

The following relationships are created/mapped:

| Source Entity \`_type\` | Relationship \`_class\` | Target Entity \`_type\` |
| --------------------- | --------------------- | --------------------- |
| \`the_root\`            | **HAS**               | \`my_account\`          |

<!--
********************************************************************************
END OF GENERATED DOCUMENTATION AFTER BELOW MARKER
********************************************************************************
-->
<!-- {J1_DOCUMENTATION_MARKER_END} -->"
`;

exports[`document loads the integration and writes documentation when documentation markers already exist 1`] = `
"# Integration with JupiterOne

## Setup

In this section, please provide details about how to set up the integration with
JupiterOne. This may require provisioning some resources on the provider's side
(perhaps a role, app, or api key) and passing information over to JupiterOne.

<!-- {J1_DOCUMENTATION_MARKER_START} -->
<!--
********************************************************************************
NOTE: ALL OF THE FOLLOWING DOCUMENTATION IS GENERATED USING THE
\\"j1-integration document\\" COMMAND. DO NOT EDIT BY HAND! PLEASE SEE THE DEVELOPER
DOCUMENTATION FOR USAGE INFORMATION:

https://github.com/JupiterOne/sdk/blob/main/docs/integrations/development.md
********************************************************************************
-->

## Data Model

### Entities

The following entities are created:

| Resources | Entity \`_type\` | Entity \`_class\` |
| --------- | -------------- | --------------- |
| The Group | \`my_group\`     | \`Group\`         |
| The User  | \`my_user\`      | \`User\`          |

### Relationships

The following relationships are created/mapped:

| Source Entity \`_type\` | Relationship \`_class\` | Target Entity \`_type\` |
| --------------------- | --------------------- | --------------------- |
| \`my_group\`            | **HAS**               | \`my_user\`             |

<!--
********************************************************************************
END OF GENERATED DOCUMENTATION AFTER BELOW MARKER
********************************************************************************
-->
<!-- {J1_DOCUMENTATION_MARKER_END} -->
"
`;

exports[`document loads the integration with array of entity classes and writes document 1`] = `
"# Integration with JupiterOne

## Setup

In this section, please provide details about how to set up the integration with
JupiterOne. This may require provisioning some resources on the provider's side
(perhaps a role, app, or api key) and passing information over to JupiterOne.


<!-- {J1_DOCUMENTATION_MARKER_START} -->
<!--
********************************************************************************
NOTE: ALL OF THE FOLLOWING DOCUMENTATION IS GENERATED USING THE
\\"j1-integration document\\" COMMAND. DO NOT EDIT BY HAND! PLEASE SEE THE DEVELOPER
DOCUMENTATION FOR USAGE INFORMATION:

https://github.com/JupiterOne/sdk/blob/main/docs/integrations/development.md
********************************************************************************
-->

## Data Model

### Entities

The following entities are created:

| Resources | Entity \`_type\` | Entity \`_class\`  |
| --------- | -------------- | ---------------- |
| The Group | \`my_group\`     | \`Group\`, \`Other\` |

<!--
********************************************************************************
END OF GENERATED DOCUMENTATION AFTER BELOW MARKER
********************************************************************************
-->
<!-- {J1_DOCUMENTATION_MARKER_END} -->"
`;

exports[`document loads the integration with entities and relationships and writes documentation results 1`] = `
"# Integration with JupiterOne

## Setup

In this section, please provide details about how to set up the integration with
JupiterOne. This may require provisioning some resources on the provider's side
(perhaps a role, app, or api key) and passing information over to JupiterOne.


<!-- {J1_DOCUMENTATION_MARKER_START} -->
<!--
********************************************************************************
NOTE: ALL OF THE FOLLOWING DOCUMENTATION IS GENERATED USING THE
\\"j1-integration document\\" COMMAND. DO NOT EDIT BY HAND! PLEASE SEE THE DEVELOPER
DOCUMENTATION FOR USAGE INFORMATION:

https://github.com/JupiterOne/sdk/blob/main/docs/integrations/development.md
********************************************************************************
-->

## Data Model

### Entities

The following entities are created:

| Resources | Entity \`_type\` | Entity \`_class\` |
| --------- | -------------- | --------------- |
| The Group | \`my_group\`     | \`Group\`         |
| The User  | \`my_user\`      | \`User\`          |

### Relationships

The following relationships are created/mapped:

| Source Entity \`_type\` | Relationship \`_class\` | Target Entity \`_type\` |
| --------------------- | --------------------- | --------------------- |
| \`my_group\`            | **HAS**               | \`my_user\`             |

<!--
********************************************************************************
END OF GENERATED DOCUMENTATION AFTER BELOW MARKER
********************************************************************************
-->
<!-- {J1_DOCUMENTATION_MARKER_END} -->"
`;

exports[`document loads the integration with entity and mapped relationship and writes documentation results 1`] = `
"# Integration with JupiterOne

## Setup

In this section, please provide details about how to set up the integration with
JupiterOne. This may require provisioning some resources on the provider's side
(perhaps a role, app, or api key) and passing information over to JupiterOne.


<!-- {J1_DOCUMENTATION_MARKER_START} -->
<!--
********************************************************************************
NOTE: ALL OF THE FOLLOWING DOCUMENTATION IS GENERATED USING THE
\\"j1-integration document\\" COMMAND. DO NOT EDIT BY HAND! PLEASE SEE THE DEVELOPER
DOCUMENTATION FOR USAGE INFORMATION:

https://github.com/JupiterOne/sdk/blob/main/docs/integrations/development.md
********************************************************************************
-->

## Data Model

### Entities

The following entities are created:

| Resources | Entity \`_type\` | Entity \`_class\` |
| --------- | -------------- | --------------- |
| The Group | \`my_group\`     | \`Group\`         |

### Mapped Relationships

The following mapped relationships are created:

| Source Entity \`_type\` | Relationship \`_class\` | Target Entity \`_type\` | Direction |
| --------------------- | --------------------- | --------------------- | --------- |
| \`my_group\`            | **HAS**               | \`*your_user*\`         | FORWARD   |

<!--
********************************************************************************
END OF GENERATED DOCUMENTATION AFTER BELOW MARKER
********************************************************************************
-->
<!-- {J1_DOCUMENTATION_MARKER_END} -->"
`;

exports[`document loads the integration without entities and writes documentation results 1`] = `
"# Integration with JupiterOne

## Setup

In this section, please provide details about how to set up the integration with
JupiterOne. This may require provisioning some resources on the provider's side
(perhaps a role, app, or api key) and passing information over to JupiterOne.


<!-- {J1_DOCUMENTATION_MARKER_START} -->
<!--
********************************************************************************
NOTE: ALL OF THE FOLLOWING DOCUMENTATION IS GENERATED USING THE
\\"j1-integration document\\" COMMAND. DO NOT EDIT BY HAND! PLEASE SEE THE DEVELOPER
DOCUMENTATION FOR USAGE INFORMATION:

https://github.com/JupiterOne/sdk/blob/main/docs/integrations/development.md
********************************************************************************
-->

## Data Model

### Relationships

The following relationships are created/mapped:

| Source Entity \`_type\` | Relationship \`_class\` | Target Entity \`_type\` |
| --------------------- | --------------------- | --------------------- |
| \`the_root\`            | **HAS**               | \`my_account\`          |

<!--
********************************************************************************
END OF GENERATED DOCUMENTATION AFTER BELOW MARKER
********************************************************************************
-->
<!-- {J1_DOCUMENTATION_MARKER_END} -->"
`;

exports[`document loads the integration without relationships and writes documentation results 1`] = `
"# Integration with JupiterOne

## Setup

In this section, please provide details about how to set up the integration with
JupiterOne. This may require provisioning some resources on the provider's side
(perhaps a role, app, or api key) and passing information over to JupiterOne.


<!-- {J1_DOCUMENTATION_MARKER_START} -->
<!--
********************************************************************************
NOTE: ALL OF THE FOLLOWING DOCUMENTATION IS GENERATED USING THE
\\"j1-integration document\\" COMMAND. DO NOT EDIT BY HAND! PLEASE SEE THE DEVELOPER
DOCUMENTATION FOR USAGE INFORMATION:

https://github.com/JupiterOne/sdk/blob/main/docs/integrations/development.md
********************************************************************************
-->

## Data Model

### Entities

The following entities are created:

| Resources   | Entity \`_type\` | Entity \`_class\` |
| ----------- | -------------- | --------------- |
| The Account | \`my_account\`   | \`Account\`       |
| The Group   | \`my_groups\`    | \`Group\`         |
| The User    | \`my_user\`      | \`User\`          |

<!--
********************************************************************************
END OF GENERATED DOCUMENTATION AFTER BELOW MARKER
********************************************************************************
-->
<!-- {J1_DOCUMENTATION_MARKER_END} -->"
`;

exports[`document should allow passing a file path for the generated documentation 1`] = `
"# Integration with JupiterOne

## Setup

In this section, please provide details about how to set up the integration with
JupiterOne. This may require provisioning some resources on the provider's side
(perhaps a role, app, or api key) and passing information over to JupiterOne.


<!-- {J1_DOCUMENTATION_MARKER_START} -->
<!--
********************************************************************************
NOTE: ALL OF THE FOLLOWING DOCUMENTATION IS GENERATED USING THE
\\"j1-integration document\\" COMMAND. DO NOT EDIT BY HAND! PLEASE SEE THE DEVELOPER
DOCUMENTATION FOR USAGE INFORMATION:

https://github.com/JupiterOne/sdk/blob/main/docs/integrations/development.md
********************************************************************************
-->

## Data Model

### Entities

The following entities are created:

| Resources | Entity \`_type\` | Entity \`_class\` |
| --------- | -------------- | --------------- |
| The Group | \`my_group\`     | \`Group\`         |

<!--
********************************************************************************
END OF GENERATED DOCUMENTATION AFTER BELOW MARKER
********************************************************************************
-->
<!-- {J1_DOCUMENTATION_MARKER_END} -->"
`;
